File: framec_tests/cpp/var_scope/var_scope.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <iostream>
#include <vector>
#include <any>
#include <string>
#include <unordered_map>
using namespace std;
#include "../FrameLang/FrameLang.h"

//=============== Compartment ==============//

using FuncPtr = void(*)();

class VarScopeCompartment
{
public:
    int state;
    
    VarScopeCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class VarScope
{
private:
    VarScopeCompartment *_compartment_;
    VarScopeCompartment *_nextCompartment_;
    
    
    
public:
    VarScope()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(VarScopeState::INIT);
        
        _compartment_ = new VarScopeCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class VarScopeState
    {
        INIT = 0,
        NN = 1,
        NY = 2,
        YN = 3,
        YY = 4
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(VarScopeState::INIT))
        {
            this->_sInit_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(VarScopeState::NN))
        {
            this->_sNN_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(VarScopeState::NY))
        {
            this->_sNY_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(VarScopeState::YN))
        {
            this->_sYN_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(VarScopeState::YY))
        {
            this->_sYY_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            VarScopeCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void to_nn() {
            FrameEvent e("to_nn", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void to_ny() {
            FrameEvent e("to_ny", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void to_yn() {
            FrameEvent e("to_yn", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void to_yy() {
            FrameEvent e("to_yy", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void nn(string d) {
            unordered_map<string, any> params;
            params["d"] = d;

            
            FrameEvent e("nn", params);
            this->_mux_(&e);
        }
        
        void ny(string d) {
            unordered_map<string, any> params;
            params["d"] = d;

            
            FrameEvent e("ny", params);
            this->_mux_(&e);
        }
        
        void yn(string d,string x) {
            unordered_map<string, any> params;
            params["d"] = d;

            
            params["x"] = x;

            
            FrameEvent e("yn", params);
            this->_mux_(&e);
        }
        
        void yy(string d,string x) {
            unordered_map<string, any> params;
            params["d"] = d;

            
            params["x"] = x;

            
            FrameEvent e("yy", params);
            this->_mux_(&e);
        }
        
        void sigils(string x) {
            unordered_map<string, any> params;
            params["x"] = x;

            
            FrameEvent e("sigils", params);
            this->_mux_(&e);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sInit_(FrameEvent *e)
    {
        if (e->_message == "to_nn") {
             std::cout << "Transition to NN state." << std::endl;
            VarScopeCompartment *compartment =  new VarScopeCompartment(static_cast<int>(VarScopeState::NN));
            compartment->stateArgs["b"] = std::string("$NN[b]");
            compartment->stateVars["c"] = std::string("$NN.c");
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_ny") {
            VarScopeCompartment *compartment =  new VarScopeCompartment(static_cast<int>(VarScopeState::NY));
             compartment->stateArgs["b"] = std::string("$NY[b]");
            compartment->stateVars["c"] = std::string("$NY.c");
            compartment->stateVars["x"] = std::string("$NY.x");
            
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_yn") {
            VarScopeCompartment *compartment =  new VarScopeCompartment(static_cast<int>(VarScopeState::YN));
            compartment->stateArgs["b"] = std::string("$YN[b]");
            compartment->stateArgs["x"] = std::string("$YN[x]");
            compartment->stateVars["c"] = std::string("$YN.c");
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "to_yy") {
            VarScopeCompartment *compartment =  new VarScopeCompartment(static_cast<int>(VarScopeState::YY));
            compartment->stateArgs["b"] = std::string("$YY[b]");
            compartment->stateArgs["x"] = std::string("$YY[x]");
            compartment->stateVars["c"] = std::string("$YY.c");
            compartment->stateVars["x"] = std::string("$YY.x");
            
            this->_transition_(compartment);
            return;
        }
    }
    
    void _sNN_(FrameEvent *e)
    {
        if (e->_message == "nn") {
            string et  = "|nn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(this->x);
            return;
        }
        else if (e->_message == "ny") {
            string et  = "|ny|.e";
            string x  = "|ny|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "yn") {
            string et  = "|yn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(any_cast<string>(e->_parameters["x"]));
            return;
        }
        else if (e->_message == "yy") {
            string et  = "|yy|.e";
            string x  = "|yy|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "sigils") {
            log_do(this->x);
            return;
        }
    }  //  var x:string = "|sigils|.x"
  //  log(||[x])
  //  log(||.x)

    
    void _sNY_(FrameEvent *e)
    {
        if (e->_message == "nn") {
            string et  = "|nn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do((any_cast<string>(this->_compartment_->stateVars["x"])));
            return;
        }
        else if (e->_message == "ny") {
            string et  = "|ny|.e";
            string x  = "|ny|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "yn") {
            string et  = "|yn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(any_cast<string>(e->_parameters["x"]));
            return;
        }
        else if (e->_message == "yy") {
            string et  = "|yy|.e";
            string x  = "|yy|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "sigils") {
            log_do(this->x);
            return;
        }
    }  //  var x:string = "|sigils|.x"
  //  log($.x)
  //  log(||[x])
  //  log(||.x)

    
    void _sYN_(FrameEvent *e)
    {
        if (e->_message == "nn") {
            string et  = "|nn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(any_cast<string>(this->_compartment_->stateArgs["x"]));
            return;
        }
        else if (e->_message == "ny") {
            string et  = "|ny|.e";
            string x  = "|ny|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "yn") {
            string et  = "|yn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(any_cast<string>(e->_parameters["x"]));
            return;
        }
        else if (e->_message == "yy") {
            string et  = "|yy|.e";
            string x  = "|yy|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "sigils") {
            log_do(this->x);
            return;
        }
    }  //  var x:string = "|sigils|.x"
  //  log($[x])
  //  log(||[x])
  //  log(||.x)

    
    void _sYY_(FrameEvent *e)
    {
        if (e->_message == "nn") {
            string et  = "|nn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do((any_cast<string>(this->_compartment_->stateVars["x"])));
            return;
        }
        else if (e->_message == "ny") {
            string et  = "|ny|.e";
            string x  = "|ny|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "yn") {
            string et  = "|yn|.e";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(any_cast<string>(e->_parameters["x"]));
            return;
        }
        else if (e->_message == "yy") {
            string et  = "|yy|.e";
            string x  = "|yy|.x";
            log_do(this->a);
            log_do(any_cast<string>(this->_compartment_->stateArgs["b"]));
            log_do((any_cast<string>(this->_compartment_->stateVars["c"])));
            log_do(any_cast<string>(e->_parameters["d"]));
            log_do(et);
            log_do(x);
            return;
        }
        else if (e->_message == "sigils") {
            log_do(this->x);
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    
public:
    void log_do(string s)
    {
         tape.push_back(s);
    }
    
    // Unimplemented Actions
    public:
    
    //===================== Domain Block ===================//
    
    
    public:
    string a  = "#.a";
    public:
    string x  = "#.x";
    public:
    std::vector<std::string> tape ;
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(VarScopeCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(VarScopeCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class VarScopeController : public VarScope
{
public:
	VarScopeController() : VarScope() {}
};

********************/

File: framec_tests/golang/var_scope/var_scope.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files


func NewVarScope() VarScope {
    m := &varScopeStruct{}
    
    // Validate interfaces
    var _ VarScope = m
    var _ VarScope_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewVarScopeCompartment(VarScopeState_Init)
    
    // Override domain variables.
    m.a = "#.a"
    m.x = "#.x"
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type VarScopeState uint

const (
    VarScopeState_Init VarScopeState = iota
    VarScopeState_NN
    VarScopeState_NY
    VarScopeState_YN
    VarScopeState_YY
)

type VarScope interface {
    To_nn() 
    To_ny() 
    To_yn() 
    To_yy() 
    Nn(d string) 
    Ny(d string) 
    Yn(d string,x string) 
    Yy(d string,x string) 
    Sigils(x string) 
}

type VarScope_actions interface {
    log(s string) 
}


type varScopeStruct struct {
    _compartment_ *VarScopeCompartment
    _nextCompartment_ *VarScopeCompartment
    a string
    x string
    tape []string
}

//===================== Interface Block ===================//

func (m *varScopeStruct) To_nn()  {
    e := framelang.FrameEvent{Msg:"to_nn"}
    m._mux_(&e)
}

func (m *varScopeStruct) To_ny()  {
    e := framelang.FrameEvent{Msg:"to_ny"}
    m._mux_(&e)
}

func (m *varScopeStruct) To_yn()  {
    e := framelang.FrameEvent{Msg:"to_yn"}
    m._mux_(&e)
}

func (m *varScopeStruct) To_yy()  {
    e := framelang.FrameEvent{Msg:"to_yy"}
    m._mux_(&e)
}

func (m *varScopeStruct) Nn(d string)  {
    params := make(map[string]interface{})
    params["d"] = d
    e := framelang.FrameEvent{Msg:"nn", Params:params}
    m._mux_(&e)
}

func (m *varScopeStruct) Ny(d string)  {
    params := make(map[string]interface{})
    params["d"] = d
    e := framelang.FrameEvent{Msg:"ny", Params:params}
    m._mux_(&e)
}

func (m *varScopeStruct) Yn(d string,x string)  {
    params := make(map[string]interface{})
    params["d"] = d
    params["x"] = x
    e := framelang.FrameEvent{Msg:"yn", Params:params}
    m._mux_(&e)
}

func (m *varScopeStruct) Yy(d string,x string)  {
    params := make(map[string]interface{})
    params["d"] = d
    params["x"] = x
    e := framelang.FrameEvent{Msg:"yy", Params:params}
    m._mux_(&e)
}

func (m *varScopeStruct) Sigils(x string)  {
    params := make(map[string]interface{})
    params["x"] = x
    e := framelang.FrameEvent{Msg:"sigils", Params:params}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *varScopeStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case VarScopeState_Init:
        m._VarScopeState_Init_(e)
    case VarScopeState_NN:
        m._VarScopeState_NN_(e)
    case VarScopeState_NY:
        m._VarScopeState_NY_(e)
    case VarScopeState_YN:
        m._VarScopeState_YN_(e)
    case VarScopeState_YY:
        m._VarScopeState_YY_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *varScopeStruct) _VarScopeState_Init_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "to_nn":
        compartment := NewVarScopeCompartment(VarScopeState_NN)
        compartment.StateArgs["b"] = "$NN[b]"
        
        compartment.StateVars["c"] = "$NN.c"
        
        m._transition_(compartment)
        
        return
    case "to_ny":
        compartment := NewVarScopeCompartment(VarScopeState_NY)
        compartment.StateArgs["b"] = "$NY[b]"
        
        compartment.StateVars["c"] = "$NY.c"
        
        compartment.StateVars["x"] = "$NY.x"
        
        m._transition_(compartment)
        
        return
    case "to_yn":
        compartment := NewVarScopeCompartment(VarScopeState_YN)
        compartment.StateArgs["b"] = "$YN[b]"
        
        compartment.StateArgs["x"] = "$YN[x]"
        
        compartment.StateVars["c"] = "$YN.c"
        
        m._transition_(compartment)
        
        return
    case "to_yy":
        compartment := NewVarScopeCompartment(VarScopeState_YY)
        compartment.StateArgs["b"] = "$YY[b]"
        
        compartment.StateArgs["x"] = "$YY[x]"
        
        compartment.StateVars["c"] = "$YY.c"
        
        compartment.StateVars["x"] = "$YY.x"
        
        m._transition_(compartment)
        
        return
    }
}

func (m *varScopeStruct) _VarScopeState_NN_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "nn":
        var et string = "|nn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(m.x)
        
        return
    case "ny":
        var et string = "|ny|.e"
        var x string = "|ny|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "yn":
        var et string = "|yn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(e.Params["x"].(string))
        
        return
    case "yy":
        var et string = "|yy|.e"
        var x string = "|yy|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "sigils":
        m.log(m.x)
        
        return
    }
}  //  var x:string = "|sigils|.x"
  //  log(||[x])
  //  log(||.x)


func (m *varScopeStruct) _VarScopeState_NY_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "nn":
        var et string = "|nn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log((m._compartment_.StateVars["x"].(string)))
        
        return
    case "ny":
        var et string = "|ny|.e"
        var x string = "|ny|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "yn":
        var et string = "|yn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(e.Params["x"].(string))
        
        return
    case "yy":
        var et string = "|yy|.e"
        var x string = "|yy|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "sigils":
        m.log(m.x)
        
        return
    }
}  //  var x:string = "|sigils|.x"
  //  log($.x)
  //  log(||[x])
  //  log(||.x)


func (m *varScopeStruct) _VarScopeState_YN_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "nn":
        var et string = "|nn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log((m._compartment_.StateArgs["x"].(string)))
        
        return
    case "ny":
        var et string = "|ny|.e"
        var x string = "|ny|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "yn":
        var et string = "|yn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(e.Params["x"].(string))
        
        return
    case "yy":
        var et string = "|yy|.e"
        var x string = "|yy|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "sigils":
        m.log(m.x)
        
        return
    }
}  //  var x:string = "|sigils|.x"
  //  log($[x])
  //  log(||[x])
  //  log(||.x)


func (m *varScopeStruct) _VarScopeState_YY_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "nn":
        var et string = "|nn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log((m._compartment_.StateVars["x"].(string)))
        
        return
    case "ny":
        var et string = "|ny|.e"
        var x string = "|ny|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "yn":
        var et string = "|yn|.e"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(e.Params["x"].(string))
        
        return
    case "yy":
        var et string = "|yy|.e"
        var x string = "|yy|.x"
        m.log(m.a)
        m.log((m._compartment_.StateArgs["b"].(string)))
        m.log((m._compartment_.StateVars["c"].(string)))
        m.log(e.Params["d"].(string))
        m.log(et)
        m.log(x)
        
        return
    case "sigils":
        m.log(m.x)
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *varScopeStruct) _transition_(compartment *VarScopeCompartment) {
    m._nextCompartment_ = compartment
}

func (m *varScopeStruct) _do_transition_(nextCompartment *VarScopeCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *varScopeStruct) log(s string)  {}

********************************************************/

//=============== Compartment ==============//

type VarScopeCompartment struct {
    State VarScopeState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewVarScopeCompartment(state VarScopeState) *VarScopeCompartment {
    c := &VarScopeCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}